### DISCUSSION TITLE
EmbedMolecule() is no longer random?

### DISCUSSION THREAD

**Discussion by jepdavidson** [rdkit/rdkit#5400] [Q&A]
Hi,

I need some help here to figure-out what I'm doing wrong(!).  I am finding that calling `AllChem.EmbedMolecule(m)` with the default randomSeed value of -1 behaves as if -1 is like any other static seed.  That is, the atom positions of the generated conformer appear to be reproducible.

Here's a quick example that just prints the coordinates of the first atom:
```python
from rdkit import Chem
from rdkit.Chem import AllChem
from rdkit.rdBase import rdkitVersion


print(rdkitVersion)
m = Chem.MolFromSmiles("CNCCOCCOC")
mh = Chem.AddHs(m)
AllChem.EmbedMolecule(mh)
conf = mh.GetConformer(0)
posns = conf.GetPositions()
print(posns[0])
```

And here's the output I get if I run multiple times (and it is the same if I explicitly set `randomSeed=-1`): 

2022.03.3
[-3.68242354  0.44139289  0.42039143]

If I manually set the randomSeed to something else (eg 123) I see different coordinates as expected:

2022.03.3
[ 4.16002579 -0.67806193 -0.07700259]

Did something happen to the random number generator, or am I missing something obvious?!

Kind regards

James

**Comment by ptosco** [comment]
@jepdavidson Hi James, what I would expect to get is: the same, deterministic sequence of conformers depending on the number I seed the generator with.
So, for example:
1. I seed the generator calling `EmbedMolecule` with `seed=1234`, and I get a confomer
2. I  call `EmbedMolecule` three times in a row without passing the `seed`parameter and I get 3 more conformers
3. I repeat steps (1) and (2): I would expect to find the same 4 conformers as before. 
4. I repeat steps (1) and (2) with `seed=5678` and I get 4 different conformers. 

This is not what I get.
Only when I seed the generator afresh I get the same conformers, as expected. However, when I call it a 2nd time, I always get a different conformer.
Looking at the code in `Code/GraphMol/DistGeomHelpers/Embedder.cpp`
https://github.com/rdkit/rdkit/blob/a88a89ec34d1bb69359ab74472d0023965f26592/Code/GraphMol/DistGeomHelpers/Embedder.cpp#L701

One may see that when `seed=-1` the random number is always taken from a static generator which was initialized with the seed 42, independently of the seed that was previously used to seed `EmbedMolecule`.
This means that currently it is not actually possible to get a deterministic sequence of conformers given an initial random seed, as only the first conformer drawn after setting the seed honours the new seed, while the subsequent ones keep drawing from the static generator originally seeded with 42.

See below a short Pythons script that exemplifies what I mean:
```
# random_seed.py
import sys
from rdkit import Chem
from rdkit.Chem import rdDistGeom

m = Chem.MolFromSmiles("OCCCCCCN")
mh = Chem.AddHs(m)

if len(sys.argv) > 1:
    seed = int(sys.argv[1])
else:
    seed = -1
i = 0
i += 1
print("Do not seed generator")
rdDistGeom.EmbedMolecule(mh)
print(i, tuple(mh.GetConformer().GetAtomPosition(0)))

for j in range(3):
    i += 1
    rdDistGeom.EmbedMolecule(mh)
    print(i, tuple(mh.GetConformer().GetAtomPosition(0)))

i += 1
print(f"Seed generator with {seed}")
rdDistGeom.EmbedMolecule(mh, randomSeed=seed)
print(i, tuple(mh.GetConformer().GetAtomPosition(0)))

for j in range(3):
    i += 1
    rdDistGeom.EmbedMolecule(mh)
    print(i, tuple(mh.GetConformer().GetAtomPosition(0))) 
```
When I run without a seed (i.e., `seed=-1`), I get
```
$ python random_seed.py
Do not seed generator
1 (3.796601034565138, -0.22157620334874803, 0.19256860495294528)
2 (2.688976683041885, -0.13153317760325686, -1.5641934953375132)
3 (-3.0113832383063537, 1.4207055127162003, -0.09169402622957816)
4 (-3.2813402531700975, -1.3976004337863945, -0.5219887975079607)
Seed generator with -1
5 (2.542863140750699, 1.0336738944664703, 1.049357414005483)
6 (-3.3992889539535533, -0.4882793979515009, 0.7898011122929305)
7 (-3.9240945704047725, -1.1273449323782287, -0.5410651075501627)
8 (-3.252874615056809, -0.5904438179817006, -0.43623887663026034)
Seed generator with -1
9 (-3.6990070497071574, 1.4658529319215388, -0.2353557181556197)
10 (3.4570511722404915, 1.289475223183645, 0.06105083531690132)
11 (3.32026003961218, -0.20083186881659038, 1.314278887985992)
12 (-2.5662427157873, 0.8572077915491663, -0.9840824701610305) 
```
This is the expected behaviour: I get 12 different conformers as the generator is seeded at startup with 42, and subsequent random numbers are drawn from a deterministic distribution.
If I run the script again, as expected I get the same conformers:
```
$ python random_seed.py 
Do not seed generator
1 (3.796601034565138, -0.22157620334874803, 0.19256860495294528)
2 (2.688976683041885, -0.13153317760325686, -1.5641934953375132)
3 (-3.0113832383063537, 1.4207055127162003, -0.09169402622957816)
4 (-3.2813402531700975, -1.3976004337863945, -0.5219887975079607)
Seed generator with -1
5 (2.542863140750699, 1.0336738944664703, 1.049357414005483)
6 (-3.3992889539535533, -0.4882793979515009, 0.7898011122929305)
7 (-3.9240945704047725, -1.1273449323782287, -0.5410651075501627)
8 (-3.252874615056809, -0.5904438179817006, -0.43623887663026034)
Seed generator with -1
9 (-3.6990070497071574, 1.4658529319215388, -0.2353557181556197)
10 (3.4570511722404915, 1.289475223183645, 0.06105083531690132)
11 (3.32026003961218, -0.20083186881659038, 1.314278887985992)
12 (-2.5662427157873, 0.8572077915491663, -0.9840824701610305)
```
The unexpected comes when I use 1234 as seed:
```

Do not seed generator
1 (3.796601034565138, -0.22157620334874803, 0.19256860495294528)
2 (2.688976683041885, -0.13153317760325686, -1.5641934953375132)
3 (-3.0113832383063537, 1.4207055127162003, -0.09169402622957816)
4 (-3.2813402531700975, -1.3976004337863945, -0.5219887975079607)
Seed generator with 1234
5 (4.3688209692078654, -0.2597722967813734, -0.6699759158153946)
6 (2.542863140750699, 1.0336738944664703, 1.049357414005483)
7 (-3.3992889539535533, -0.4882793979515009, 0.7898011122929305)
8 (-3.9240945704047725, -1.1273449323782287, -0.5410651075501627)
Seed generator with 1234
9 (4.3688209692078654, -0.2597722967813734, -0.6699759158153946)
10 (-3.252874615056809, -0.5904438179817006, -0.43623887663026034)
11 (-3.6990070497071574, 1.4658529319215388, -0.2353557181556197)
12 (3.4570511722404915, 1.289475223183645, 0.06105083531690132)
```
The first 4 conformers are the same as before, which is correct as I am drawing from the generator initialized with 42.
Then I seed with 1234: I would expect 4 different conformers.
Instead, only conformer 5 is different, while 6-8 are the same as 5-7 from the previous run of the script, because the unseeded draws keep drawing from the 42 distribution, while they should in my opinion draw from the 1234 distribution.
Draw 9 is the same as 5, which is correct, while draws 10-12 are the same as 8-10 from the previous run, for the mentioned reasons, while they should be identical to 5-8 from this run.

In summary, I believe the current behaviour is incorrect, as the correct one should be the one I described above.
@greglandrum do you agree with me? If so, I will submit a PR to fix it based on my description of the intended behaviour.

**Reply by jepdavidson** [reply]
Hi @ptosco ,

Thanks for the very detailed response.  I think I understand the issue that you have raised (let's wait to see what @greglandrum says!).

I think my issue is my reading of the docstring:

```
randomSeed : provide a seed for the random number generator
so that the same coordinates can be obtained for a molecule on multiple runs. If -1, the RNG will not be seeded.
```

I always thought **the RNG will not be seeded** meant that the outcome would be 'random' (and not reproducible) for a single embedding...  I need to reassess my life!

Kind regards

James

**Reply by ptosco** [reply]
@jepdavidson If you want a non-reproducible conformer at each call, you can already accomplish that with the current code; just do something like:
```
def time_rand():
    t = time.time()
    t -= int(t)
    return int(t * 1.e7)

rdDistGeom.EmbedMolecule(mh, randomSeed=time_rand())
print(tuple(mh.GetConformer().GetAtomPosition(0)))
```
since the conformer randomness that you get when you set the seed is guaranteed to depend on the seed, and if you seed from the current time you will get a non-reproducible conformer at each call.

**Reply by greglandrum** [reply]
Just to put something in here quickly... I think I disagree with what @ptosco is saying, but I also think I can't reproduce the problem that @jepdavidson is having. So I'm confused and need to dig more deeply.

To try and remove some variables: @jepdavidson : which version of the RDKit are you using and on which OS?

**Reply by greglandrum** [reply]
Here's a very simple demo session showing what I get when generating a single conformer with the most recent RDKit (though nothing here should have changed recently) on windows:
```
In [48]: print(rdkit.__version__)
2022.03.4

In [49]: m = Chem.AddHs(Chem.MolFromSmiles('FCCO'))

In [50]: rdDistGeom.EmbedMolecule(m)
Out[50]: 0

In [51]: m.GetConformer().GetPositions()[0]
Out[51]: array([-1.23148385,  0.28205554,  1.20194215])

In [52]: rdDistGeom.EmbedMolecule(m)
Out[52]: 0

In [53]: m.GetConformer().GetPositions()[0]
Out[53]: array([ 0.89385681,  1.37269531, -0.58334134])

In [54]: rdDistGeom.EmbedMolecule(m,randomSeed=-1)
Out[54]: 0

In [55]: m.GetConformer().GetPositions()[0]
Out[55]: array([-1.41330691,  0.91243551,  0.71284387])

In [56]: rdDistGeom.EmbedMolecule(m,randomSeed=-1)
Out[56]: 0

In [57]: m.GetConformer().GetPositions()[0]
Out[57]: array([-0.95603252,  1.40092002,  0.50900284])
```
There I get a different conformer every time, which is what I'd expect when not providing a random seed or when I set it to -1 (which is the default anyway)

If I specify a seed I get something reproducible:
```
In [58]: rdDistGeom.EmbedMolecule(m,randomSeed=0xf00d)
Out[58]: 0

In [59]: m.GetConformer().GetPositions()[0]
Out[59]: array([-0.9195272 ,  1.14976836,  0.9067938 ])

In [60]: rdDistGeom.EmbedMolecule(m,randomSeed=0xf00d)
Out[60]: 0

In [61]: m.GetConformer().GetPositions()[0]
Out[61]: array([-0.9195272 ,  1.14976836,  0.9067938 ])
```

**Reply by greglandrum** [reply]
I get similar, expected, behavior when I ask to have multiple conformers generated.

Providing a seed gives reproducible results:
```
In [62]: rdDistGeom.EmbedMultipleConfs(m,2,randomSeed=0xf00d)
Out[62]: <rdkit.rdBase._vectint at 0x2ad18020ac0>

In [63]: m.GetConformer().GetPositions()[0]
Out[63]: array([-0.9195272 ,  1.14976836,  0.9067938 ])

In [64]: m.GetConformer(1).GetPositions()[0]
Out[64]: array([ 1.44597299, -0.07003134,  1.21117145])

In [65]: rdDistGeom.EmbedMultipleConfs(m,2,randomSeed=0xf00d)
Out[65]: <rdkit.rdBase._vectint at 0x2ad169fb9e0>

In [66]: m.GetConformer().GetPositions()[0]
Out[66]: array([-0.9195272 ,  1.14976836,  0.9067938 ])

In [67]: m.GetConformer(1).GetPositions()[0]
Out[67]: array([ 1.44597299, -0.07003134,  1.21117145])
```

and either not providing a seed, or providing -1 as the seed, gives variable results:
```
In [68]: rdDistGeom.EmbedMultipleConfs(m,2,randomSeed=-1)
Out[68]: <rdkit.rdBase._vectint at 0x2ad18359ac0>

In [69]: m.GetConformer().GetPositions()[0]
Out[69]: array([-1.52719116, -1.23659547,  0.25369273])

In [70]: m.GetConformer(1).GetPositions()[0]
Out[70]: array([-1.03209435,  0.68892757,  1.22116939])

In [71]: rdDistGeom.EmbedMultipleConfs(m,2,randomSeed=-1)
Out[71]: <rdkit.rdBase._vectint at 0x2ad1620bc10>

In [72]: m.GetConformer().GetPositions()[0]
Out[72]: array([ 2.00282095, -0.34954266, -0.22929214])

In [73]: m.GetConformer(1).GetPositions()[0]
Out[73]: array([ 1.09994052,  0.93307461, -1.12133684])
```

**Reply by greglandrum** [reply]
And now, after looking at my results, and re-reading @jepdavidson's question a third or fourth time (it takes me a while sometimes...) I understand what the problem is: you get the same sequence of "random" conformers every time you run python.

@ptosco pointed out the problem here: there's a global distribution which is (incorrectly) seeded with 42 every time. We should fix that.

As @ptosco also pointed out: the workaround with the current code is to provide your own seed based on the current time.

**Reply by ptosco** [reply]
I have discussed this with @greglandrum in a lively chat offline.. I think it would be clearer and more useful to have a stateful RNG that can be seeded and then used to draw a seed-dependent series of conformers when calling `EmbedMolecule` multiple times. I also think that the `EmbedMolecule` docs for the `seed` parameter suggest that should be the behavior ("if negative, the RNG will not be seeded"). But I lost the argument, so I am leaving this note as a testament to posterity :)

**Reply by greglandrum** [reply]
Yes, we've agreed to disagree. :-)

**Reply by jepdavidson** [reply]
@greglandrum - thanks for the follow-up.
I hadn't actually put 2-and-2 together until your reply; I guess the biggest thing that changed for me was running the test as a script (with `python script.py`) rather than interactively!

Just for completeness (although now unnecessary), I had included the RDKit version in my original post (2022.03.3), but had not specified the platform...  I was running in Ubuntu20.04 via WSL2 (rdkit from conda-forge: py310h1c297d8_0)

**Reply by ptosco** [reply]
Summarizing here an answer to @jepdavidson  and also an answer to my own concerns.
* Without changing the current RDKit code, a random conformer can be obtained every time the script is called, and every time `EmbedMolecule()` is called, by seeding with current time.
* Every time the script is run, a reproducible sequence of deterministic, seed-dependent conformers can be obtained by seeding with a sequence of random ints generated with an independently seeded RNG.
Code example below:
```
# random_seed.py

import sys
import random
import time
from rdkit import Chem
from rdkit.Chem import rdDistGeom

m = Chem.MolFromSmiles("OCCCCCCN")
mh = Chem.AddHs(m)

def time_rand():
    t = time.time()
    t -= int(t)
    return int(t * 1.e7)

if len(sys.argv) > 1:
    seed = int(sys.argv[1])
    random.seed(seed)
    get_seed = lambda: random.randint(0, 1 << 31)
else:
    seed = "time"
    get_seed = time_rand

for i in range(4):
    print(f"Seed generator with {seed}")
    rdDistGeom.EmbedMolecule(mh, randomSeed=get_seed())
    print(i + 1, tuple(mh.GetConformer().GetAtomPosition(0)))
```
If I do not provide a seed as command line parameter, conformers are random at each run of the script and at each call of `EmbedMolecule()`:
```
$ python random_seed.py
Seed generator with time
1 (-2.74855896092562, 1.6255398965622188, 0.1662723343558729)
Seed generator with time
2 (3.0974101102748177, 1.5663387319272377, -0.6501502015141352)
Seed generator with time
3 (-3.7241914785111168, 1.1144953432212077, 0.005896140972008175)
Seed generator with time
4 (2.4164813477257443, 1.679102176435764, -0.7175473062637352)
$ python random_seed.py
Seed generator with time
1 (2.55619471614514, 0.9208633291191229, -1.0923195117988787)
Seed generator with time
2 (-3.462278289087107, 1.2421589363694285, -0.07095905643896688)
Seed generator with time
3 (3.2333500990834088, 1.516835144723658, 0.21009181296276477)
Seed generator with time
4 (4.042362806522934, 0.4913371161176487, -0.25251914289786376)
```
If I provide a seed as command line parameter, conformers are the same at each run of the script (given the same seed), and different but pre-determined by the seed at each call of `EmbedMolecule()`:
```
$ python random_seed.py 1234
Seed generator with 1234
1 (2.434337748349493, 1.068446703939622, 0.9688945270485905)
Seed generator with 1234
2 (-3.7170603662153434, -0.6467254300158275, -0.7457765674434248)
Seed generator with 1234
3 (-3.158685321604218, 0.9953079343897496, 1.6650501973147527)
Seed generator with 1234
4 (3.6147188955148795, -0.8863867244541342, 0.6633132697058437)
$ python random_seed.py 1234
Seed generator with 1234
1 (2.434337748349493, 1.068446703939622, 0.9688945270485905)
Seed generator with 1234
2 (-3.7170603662153434, -0.6467254300158275, -0.7457765674434248)
Seed generator with 1234
3 (-3.158685321604218, 0.9953079343897496, 1.6650501973147527)
Seed generator with 1234
4 (3.6147188955148795, -0.8863867244541342, 0.6633132697058437)
$ python random_seed.py 5678
Seed generator with 5678
1 (3.043480018917979, -1.0574386202646493, -0.3954541282264613)
Seed generator with 5678
2 (3.6801590590787767, 0.9663151875308675, 0.9205255189660813)
Seed generator with 5678
3 (3.262590155969196, -0.03726035819338493, -0.9378580127409104)
Seed generator with 5678
4 (-4.2278808373033705, 0.863396946389663, 0.37969946441396507)
$ python random_seed.py 5678
Seed generator with 5678
1 (3.043480018917979, -1.0574386202646493, -0.3954541282264613)
Seed generator with 5678
2 (3.6801590590787767, 0.9663151875308675, 0.9205255189660813)
Seed generator with 5678
3 (3.262590155969196, -0.03726035819338493, -0.9378580127409104)
Seed generator with 5678
4 (-4.2278808373033705, 0.863396946389663, 0.37969946441396507)
```
This should satisfy all use cases described above without requiring code changes.
